kvcoord: track a write buffer transformation for all requests

miraradeva opened the pull request:
Previously, the transaction write buffer kept a transformation for many requests: truly transformed requests (e.g. CPut), and other requests, like Get, which were not actually transformed but needed to be tracked in order to correctly be able to read your own writes. Scan requests, specifically, were not tracked as transformations, but should be, just like Gets (#146103).

This commit tracks all requests as transformations regardless of whether they were actually transformed. For most requests, this is needed for correctness; for requests like QueryLocksRequest and LeaseInfoRequest, it is not necessary, but it is harmless and simplifies the code.

Fixes: [#146103](https://github.com/cockroachdb/cockroach/issues/146103)

Release note: None

miraradeva committed:
kvcoord: buffered writes don't correctly handle read-your-own-writes for scans

This commit expands the `TestTxnBufferedWriteReadYourOwnWrites` test with
Scan requests. As part of that, it repros the issue described
in #146103.

Informs: #146103

Release note: None

miraradeva committed:
kvcoord: track a write buffer transformation for all requests

Previously, the transaction write buffer kept a transformation for many
requests: truly transformed requests (e.g. CPut), and other requests,
like Get, which were not actually transformed but needed to be tracked
in order to correctly be able to read your own writes. Scan requests,
specifically, were not tracked as transformations, but should be, just
like Gets (#146103).

This commit tracks all requests as transformations regardless of whether
they were actually transformed. For most requests, this is needed for
correctness; for requests like QueryLocksRequest and LeaseInfoRequest,
it is not necessary, but it is harmless and simplifies the code.

Fixes: #146103

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/146112)
<!-- Reviewable:end -->

yuzefovich reviewed (approved):
Nice find! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Should we backport this? SQL will never construct a batch like this, but it might still be good to fix it on 25.2.

> Reviewed 1 of 1 files at r1, 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146112#-:-OPae00K81jXFRGRl6cu:bs48igr)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @arulajmani and @stevendanna)

___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go` line 675 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/146112#-OPadWpW6eNy33x5UV48:-OPadWpX6qnZjbbrTgZz:brybzd1) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5d109059a259697ccf14bdb13c916ae948b40fc/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go#L675)):*
> ```Go
> 	baRemote.Requests = nil
> 
> 	var ts transformations
> ```

nit: we can precisely allocate this slice now.
___
*[`pkg/kv/kvclient/kvcoord/txn_test.go` line 45 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/146112#-OPab9Fb24BV58o4D7xL:-OPab9Fb24BV58o4D7xM:bu66wn0) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e2699234184fa596558d64311e5e89836950086/pkg/kv/kvclient/kvcoord/txn_test.go#L45)):*
> ```Go
> )
> 
> func checkResults(t *testing.T, expected map[string][]byte, results ...kv.Result) {
> ```

nit: should `results` still be a variadic argument? Looks like it's always a single `kv.Result` now. Might be even worth inlining since it's only called in a single spot now.


<!-- Sent from Reviewable.io -->

miraradeva committed:
kvcoord: rename transformation to requestRecord in the write buffer

Now that we keep tranck of all requests as transformations, the
term "transformation" doesn't make sense because some of the requests
aren't actually transformed.

This commit renames all transformation-related terms and
replaces "transformation" with "requestRecord", while also adding a new
`transformed` boolean field in `requestRecord` to indicate if a request
was actually transformed.

Part of: #146103

Release note: None

yuzefovich reviewed (approved):
Thanks!

> Reviewed 2 of 2 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146112#-:-OPbJb9C60E-B1zXfLKo:b-ucfknh)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @arulajmani and @stevendanna)

___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go` line 647 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/146112#-OPbIfJ6EbyfIMu9wl6t:-OPbIfJ6EbyfIMu9wl6u:b-s2yl0g) ([raw file](https://github.com/cockroachdb/cockroach/blob/98907b6b96da81684443c1a4b2b05d3fb8b49418/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go#L647)):*
> ```Go
> // Some examples of transformations include:
> //
> // 1. Blind writes (Put/Delete requests) are buffered locally. When they the
> ```

nit: `s/they the/the/`.
___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go` line 1080 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/146112#-OPbJTcv2I3TyfxNr_wt:-OPbJTcv2I3TyfxNr_wu:b-jczxqv) ([raw file](https://github.com/cockroachdb/cockroach/blob/98907b6b96da81684443c1a4b2b05d3fb8b49418/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go#L1080)):*
> ```Go
> // toResp returns the response that should be added to the batch response as
> // a result of applying the requestRecord.
> func (t requestRecord) toResp(
> ```

nit: `s/t/r/` or `s/t/rr/`.


<!-- Sent from Reviewable.io -->

stevendanna commented on pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go:
```diff
@@ -1087,26 +1058,30 @@ func (twb *txnWriteBuffer) mergeResponseWithTransformations(
 	stripped bool
-	// index of the request in the original batch to which the transformation
+	// transformed, if true, indicates that the request was transformed into a
+	// different request to be sent to the KV layer.
```
Should we specify that this is only relevant if stripped is false?

miraradeva replied:
Yes, I guess, though it depends on how we plan to use this new `transformed` flag. 

Should a blind Put, which is stripped, set `transformed = true`? Currently, I set `transformed = true` for all CPut, Put, Del. But if we only set it when `stripped == false`, then we can ensure that `stripped == true` implies `transformed == false`. Maybe that's a useful invariant.

miraradeva replied:
I made the change ^^, and added a comment about it. I don't think it makes sense for a stripped Put to be considered transformed. @stevendanna let me know if that makes sense to you.

stevendanna replied:
I like it. At the very least it is some trail that will remind us of our intention if we get confused later on.

miraradeva commented:
bors r=yuzefovich,stevendanna
